workflow/controller/operator.go

```diff
@@ -1387,6 +1408,10 @@ func (woc *wfOperationCtx) executeTemplate(nodeName string, orgTmpl wfv1.Templat
 
 	if node != nil {
 		if node.Fulfilled() {
+			if resolvedTmpl.Synchronization != nil {
+				woc.log.Debugf("Node %s is releasing Semaphore lock", node.ID)
+				woc.controller.concurrencyMgr.Release(woc.wf, node.ID, woc.wf.Namespace, resolvedTmpl.Synchronization)
```

Comment:
Is this idempotent? It will be called during every operation after the node is fulfilled

Reply:
If this is happening every operation after it's fulfilled, then the debug message will spam. Probably this message should only emit when the lock is removed from the list of holders. Otherwise it will spam.

Reply:
yes, This function will release only once.